bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/graphics/Image.java

```diff
@@ -1408,7 +1408,39 @@ public ImageData getImageData(int zoom) {
 	} finally {
 		if (pool != null) pool.release();
 	}
-	return DPIUtil.scaleImageData (device, getImageData(100), zoom, 100);
+	ImageData imageData = getImageData(100);
+	return scaledTo (imageData, zoom, 100, DPIUtil.getScalingType(imageData));
+}
+
+private ImageData scaledTo(ImageData imageData, int targetZoom, int currentZoom, int scaleType) {
+	if (imageData == null || currentZoom == targetZoom || !device.isAutoScalable()) {
+		return imageData;
+	}
+	float scaleFactor = (float) targetZoom / (float) currentZoom;
+	int scaledWidth = Math.round (imageData.width * scaleFactor);
+	int scaledHeight = Math.round (imageData.height * scaleFactor);
+	switch (scaleType) {
+	case SWT.SMOOTH:
+		return scaleUsingSmoothScaling(imageData, scaledWidth, scaledHeight);
+	default:
+		return imageData.scaledTo(scaledWidth, scaledHeight);
+	}
+}
+
+private ImageData scaleUsingSmoothScaling(ImageData imageData, int width, int height) {
+	Image original = new Image (device, (ImageDataProvider) zoom -> imageData);
```

Comment:
why do we create a new image here and not using `this` Image?

Reply:
Wouldn't that create an endless loop? When you try to create image data at a specific zoom by painting the image into a GC, which requires requesting the image data at that specific zoom, this will likely produce a stack overflow.

Reply:
The idea was that at the Image level we can access any internal API as required (including a painting the current `ImageData` instead of using `ImageDataProvider`).